src/webui/gitutils.jl

```diff
@@ -0,0 +1,159 @@
+    else
+        # First check for organization membership, and fall back to collaborator status.
+        ismember = @gf is_member(u.forge, repo.organization.login, u.user.login)
+        something(ismember, false) ||
+            @gf is_collaborator(u.forge, repo.organization.login, repo.name, u.user.login)
+    end
+    return something(hasauth, false)
+end
+
+function isauthorized(u::User{GitLab.User}, repo::GitLab.Project)
+    repo.visibility == "private" && return false
+    forge = PROVIDERS["gitlab"].client
+    hasauth = if repo.namespace.kind == "user"
+        @gf is_collaborator(forge, repo.owner.username, repo.name, u.user.id)
+    else
+        # Same as above: group membership then collaborator check.
+        ismember = @gf is_member(forge, repo.namespace.full_path, u.user.id)
+        something(ismember, false) ||
+            @gf is_collaborator(u.forge, repo.organization.login, repo.name, u.user.id)
+    end
+    return something(hasauth, false)
+end
+
+# Get the raw Project.toml text from a repository.
+function gettoml(f::GitHubAPI, repo::GitHub.Repo, ref::AbstractString)
+    fc = @gf get_file_contents(f, repo.owner.login, repo.name, "Project.toml"; ref=ref)
+    return fc === nothing ? nothing : String(base64decode(strip(fc.content)))
+end
+
+function gettoml(::GitLabAPI, repo::GitLab.Project, ref::AbstractString)
+    forge = PROVIDERS["gitlab"].client
+    fc = @gf get_file_contents(forge, repo.id, "Project.toml"; ref=ref)
+    return fc === nothing ? nothing : String(base64decode(fc.content))
+end
+
+function getcommithash(f::GitHubAPI, repo::GitHub.Repo, ref::AbstractString)
+    commit = @gf get_commit(f, repo.owner.login, repo.name, ref)
+    return commit === nothing ? nothing : commit.sha
+end
+
+function getcommithash(f::GitLabAPI, repo::GitLab.Project, ref::AbstractString)
+    forge = PROVIDERS["gitlab"].client
+    commit = @gf get_commit(forge, repo.id, ref)
+    return commit === nothing ? nothing : commit.id
+end
+
+# Get a repo's clone URL.
+cloneurl(r::GitHub.Repo) = r.clone_url
+cloneurl(r::GitLab.Project) = r.http_url_to_repo
+
+# Get a repo's tree hash.
+function gettreesha(f::GitHubAPI, r::GitHub.Repo, ref::AbstractString)
+    branch = @gf get_branch(f, r.owner.login, r.name, ref)
+    return branch === nothing ? nothing : branch.commit.commit.tree.sha
+end
+
+function gettreesha(::GitLabAPI, r::GitLab.Project, ref::AbstractString)
+    url = cloneurl(r)
+
+    if REGISTRY[] isa Registry{GitLabAPI}
+        # For private repositories, we need to insert the token into the URL.
+        host = HTTP.URI(url).host
+        token = CONFIG["gitlab"]["token"]
+        url = replace(url, host => "oauth2:$token@$host")
+    end
+
+    return try
+        mktempdir() do dir
+            dest = joinpath(dir, r.name)
+            run(`git clone $url $dest`)
+            match(r"tree (.*)", readchomp(`git -C $dest show $ref --format=raw`))[1]
+        end
+    catch
```

Comment:
This is not great.
It catches *all* exceptions, and does not even log anything?
That means it could be catching `UndefinedVarError`s etc.

Reply:
Yeah that is not ideal, this was basically a rough draft that I forgot to make nicer.

Reply:
I've edited this part to display the backtrace.